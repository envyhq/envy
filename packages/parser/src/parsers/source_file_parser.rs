use super::{
    module_declaration_parser::ModuleDeclarationParser,
    provider_declaration_parser::ProviderDeclarationParser,
};
use crate::{
    abstract_syntax_tree::{AbstractSyntaxNode, SourceFileNode},
    parser::ParserResult,
    parsers::var_declaration_parser::VarDeclarationParser,
    Parser,
};
use nv_lexer::{
    tokens::{LexerDeclarationKeyword, LexerToken},
    LexerKeyword, LexerTokenKind, LexerVarModifierKeyword,
};
use std::sync::{Arc, Mutex};

#[derive(Debug)]
pub struct SourceFileParser;

impl SourceFileParser {
    pub fn parse(tokens: &[LexerToken]) -> (usize, Arc<AbstractSyntaxNode>) {
        let root = Arc::new(AbstractSyntaxNode::SourceFile(Arc::new(SourceFileNode {
            declarations: Mutex::new(vec![]),
        })));
        let mut tokens_iter = tokens.iter().enumerate();

        let mut processed_count = 0;

        while let Some((index, token)) = tokens_iter.next() {
            processed_count += 1;
            let sub_tokens = &tokens[index..].to_vec();
            let sub_tokens = sub_tokens.to_vec();

            let result = match token.kind.clone() {
                LexerTokenKind::Keyword(LexerKeyword::VarModifierKeyword(
                    LexerVarModifierKeyword::Pub,
                ))
                | LexerTokenKind::Keyword(LexerKeyword::DeclarationKeyword(
                    LexerDeclarationKeyword::Var,
                )) => {
                    let (count, parsed_fragment) =
                        VarDeclarationParser::parse(&sub_tokens, Arc::downgrade(&root));

                    ParserResult {
                        processed_count: count - 1,
                        ast_fragment: parsed_fragment,
                    }
                }
                LexerTokenKind::Keyword(LexerKeyword::DeclarationKeyword(keyword)) => match keyword
                {
                    LexerDeclarationKeyword::Provider => {
                        let (count, parsed_fragment) =
                            ProviderDeclarationParser::parse(&sub_tokens, Arc::downgrade(&root));

                        ParserResult {
                            processed_count: count - 1,
                            ast_fragment: parsed_fragment,
                        }
                    }
                    LexerDeclarationKeyword::Module => {
                        let (count, parsed_fragment) =
                            ModuleDeclarationParser::parse(&sub_tokens, Arc::downgrade(&root));

                        ParserResult {
                            processed_count: count - 1,
                            ast_fragment: parsed_fragment,
                        }
                    }
                    _ => {
                        continue;
                    }
                },
                _ => {
                    continue;
                }
            };

            if let Some(AbstractSyntaxNode::Declaration(declaration)) = result.ast_fragment {
                if let AbstractSyntaxNode::SourceFile(source_node) = root.as_ref() {
                    source_node
                        .declarations
                        .lock()
                        .unwrap()
                        .push(Arc::new(declaration));
                }
            }

            processed_count += result.processed_count;
            if result.processed_count > 0 {
                tokens_iter.nth(result.processed_count - 1);
            }
        }

        (processed_count, root)
    }
}
