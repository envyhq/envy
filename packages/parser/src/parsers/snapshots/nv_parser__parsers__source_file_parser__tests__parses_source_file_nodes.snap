---
source: packages/parser/src/parsers/source_file_parser.rs
expression: ast
---
SourceFile:
  path: test.nv
  declarations:
    - VarDeclaration:
        identifier:
          value: my_var
          range:
            from:
              line: 1
              column: 5
            to:
              line: 1
              column: 10
        type_value:
          value: Int
          range:
            from:
              line: 1
              column: 14
            to:
              line: 1
              column: 16
        modifier: ~
        attributes: []
    - ProviderDeclaration:
        identifier:
          value: Env
          range:
            from:
              line: 3
              column: 10
            to:
              line: 3
              column: 12
        type_value:
          value: env
          range:
            from:
              line: 3
              column: 15
            to:
              line: 3
              column: 17
        attributes: []
    - ModuleDeclaration:
        identifier: MyCoolModule
        declarations:
          - VarDeclaration:
              identifier:
                value: my_cool_var
                range:
                  from:
                    line: 6
                    column: 9
                  to:
                    line: 6
                    column: 19
              type_value:
                value: String
                range:
                  from:
                    line: 6
                    column: 23
                  to:
                    line: 6
                    column: 25
              modifier: ~
              attributes: []
          - VarDeclaration:
              identifier:
                value: my_other_var
                range:
                  from:
                    line: 7
                    column: 13
                  to:
                    line: 7
                    column: 24
              type_value:
                value: Url
                range:
                  from:
                    line: 7
                    column: 28
                  to:
                    line: 7
                    column: 30
              modifier:
                value: Pub
                range:
                  from:
                    line: 7
                    column: 5
                  to:
                    line: 7
                    column: 7
              attributes: []
